cpp/src/strings/replace/multi.cu

```diff
@@ -0,0 +1,488 @@
+
+        auto const d_repl = get_replacement_string(d_pair.second);
+        if (!d_repl.empty()) { count++; }
+
+        str_ptr += keep_size + d_tgt.size_bytes();
+      }
+    }
+
+    return count;
+  }
+
+  /**
+   * @brief Retrieve the strings for each row
+   *
+   * This will return string segments as string_index_pair objects for
+   * parts of the string that are not replaced interlaced with the
+   * appropriate replacement string where replacement targets are found.
+   *
+   * This function is called only once to produce both the string_index_pair objects
+   * and the output row size in bytes.
+   *
+   * @param idx Index of the row in d_strings
+   * @param d_offsets Offsets to identify where to store the results of the replace for this string
+   * @param d_positions The target positions found in the chars column
+   * @param d_targets_offsets The offsets to identify which target positions go with this string
+   * @param d_all_strings The output of all the produced string segments
+   * @return The size in bytes of the output string for this row
+   */
+  __device__ size_type get_strings(size_type idx,
+                                   size_type const* d_offsets,
+                                   target_pair const* d_positions,
+                                   size_type const* d_targets_offsets,
+                                   string_index_pair* d_all_strings) const
+  {
+    if (!is_valid(idx)) { return 0; }
+
+    auto const d_output       = d_all_strings + d_offsets[idx];
+    auto const d_output_count = d_offsets[idx + 1] - d_offsets[idx];
+
+    auto const d_str = get_string(idx);
+    if (d_output_count == 1) {
+      d_output[0] = string_index_pair{d_str.data(), d_str.size_bytes()};
+      return d_str.size_bytes();
+    }
+
+    auto const d_str_end         = d_str.data() + d_str.size_bytes();
+    auto const base_ptr          = get_base_ptr();  //+ delim_size - 1;
+    auto const targets_positions = cudf::device_span<target_pair const>(
+      d_positions + d_targets_offsets[idx], d_targets_offsets[idx + 1] - d_targets_offsets[idx]);
+
+    size_type output_idx  = 0;
+    size_type output_size = 0;
+    auto str_ptr          = d_str.data();
+    for (auto d_pair : targets_positions) {
+      auto const d_pos   = d_pair.first;
+      auto const d_tgt   = d_targets[d_pair.second];
+      auto const tgt_ptr = base_ptr + d_pos;
+      if (str_ptr <= tgt_ptr && tgt_ptr < d_str_end) {
+        auto const keep_size = static_cast<size_type>(thrust::distance(str_ptr, tgt_ptr));
+        if (keep_size > 0) { d_output[output_idx++] = string_index_pair{str_ptr, keep_size}; }
+        output_size += keep_size;
+
+        auto const d_repl = get_replacement_string(d_pair.second);
+        if (!d_repl.empty()) {
+          d_output[output_idx++] = string_index_pair{d_repl.data(), d_repl.size_bytes()};
+        }
+        output_size += d_repl.size_bytes();
```

Comment:
Is this still true even if `d_repl.empty()==true`? Is `d_repl.size_bytes() == 0` if it is empty? I guess my question really is if this should move inside the if block.

Reply:
The `d_repl.size_bytes() == 0` will be true if `d_repl.empty()==true`
Just trying to minimize divergent logic. Seemed harmless to add 0 in lock step with the other threads in the warp.